### PR TITLE
feat(syscall): Add some network syscalls

### PR DESCRIPTION
**Pull request by haraldh**
Implemented proxying of the following syscalls:
    
* socket
* bind
* connect
* accept
* accept4
* listen
* recvfrom

### CODE REVIEW

File: tests/bin/unix_echo.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+use std::io::{self, Read, Write};
+use std::os::unix::net::{UnixListener, UnixStream};
+
+fn main() -> io::Result<()> {
+    let listener = UnixListener::bind("/tmp/enarx_unix_echo_test_to_test")?;
```
Comment (lkatalin) [inline tests/bin/unix_echo.rs:5]: I might call this something like `/tmp/enarx_unix_echo_to_bin` and `/tmp/enarx_unix_echo_from_bin` (or `echo_from_code` or whatever sounds good) to be a bit more descriptive.

File: tests/integration_tests.rs [hunk 2]
```diff
@@ -226,6 +229,50 @@ fn echo() {
     run_test("echo", 0, input.as_slice(), input.as_slice(), None);
 }
 
+#[test]
+#[serial]
+fn unix_echo() {
+    const FILENAME_IN: &'static str = "/tmp/enarx_unix_echo_test_to_test";
+    const FILENAME_OUT: &'static str = "/tmp/enarx_unix_echo_test_from_test";
```
Comment (lkatalin) [inline tests/integration_tests.rs:236]: Same comment here as `test_to_test` is more ambiguous.

**Review by lkatalin — COMMENTED** [review]
LGTM, small comment on naming in the test but not a big deal either way.

File: internal/syscall/src/lib.rs [hunk 3]
```diff
@@ -658,4 +678,207 @@ pub trait SyscallHandler: AddressValidator + Sized {
...
+        addr: UntrustedRefMut<u8>,
+        addrlen: UntrustedRefMut<libc::socklen_t>,
+        flags: libc::c_int,
+    ) -> Result {
+        self.trace("accept4", 4);
+
+        if addr.as_ptr().is_null() {
+            return unsafe {
+                self.proxy(
+                    request!(libc::SYS_accept4 => fd, addr.as_ptr(), addrlen.as_ptr(), flags),
```
Comment (connorkuehl) [inline internal/syscall/src/lib.rs:750]: The man page suggests that if `addr` is passed in as null, then a null pointer should be passed in for `addrlen`.

**Review by connorkuehl — CHANGES_REQUESTED** [review]
Please split each syscall into a separate commit

File: internal/syscall/src/lib.rs (hunk 3 above)
Comment (haraldh) [reply internal/syscall/src/lib.rs:750]: @connorkuehl  And? I just pass on what userspace gave me and userspace will get the errors it deserves.

File: tests/bin/unix_echo.rs (hunk 1 above)
Comment (haraldh) [reply tests/bin/unix_echo.rs:5]: thanks! good idea

**Comment by haraldh** [comment]
> Please split each syscall into a separate commit

@connorkuehl as you wish, please re-review
CC @lkatalin

**Review by lkatalin — APPROVED** [review]

File: tests/integration_tests.rs (hunk 2 above)
Comment (haraldh) [reply tests/integration_tests.rs]: should be resolved

File: tests/bin/bind.c [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    };
+    socklen_t sa_len = strlen(UNIX_ABSTRACT_PATH);
+
+    int fd = socket(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0);
+
+    if (fd < 0)
+        return 1;
+
+    strcpy(sa.sun_path, UNIX_ABSTRACT_PATH);
+    sa.sun_path[0] = '\0';
```
Comment (connorkuehl) [inline tests/bin/bind.c:22]: Won't this result in the empty string being passed as the path?

File: tests/bin/bind.c [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    if (bind(fd, (struct sockaddr *)&sa, offsetof(struct sockaddr_un, sun_path) + sa_len) < 0) {
```
Comment (connorkuehl) [inline tests/bin/bind.c:24]: Should the length field be only `sa_len`?

File: tests/bin/listen.c [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+    if (fd < 0)
+        return 1;
+
+    strcpy(sa.sun_path, UNIX_ABSTRACT_PATH);
+    sa.sun_path[0] = '\0';
+
+    if (bind(fd, (struct sockaddr *)&sa, offsetof(struct sockaddr_un, sun_path) + sa_len) < 0) {
+        return 2;
+    }
```
Comment (connorkuehl) [inline tests/bin/listen.c:26]: Same questions as `bind` commit.

**Review by connorkuehl — APPROVED** [review]
A couple of questions inline, but otherwise LGTM.

File: tests/bin/bind.c (hunk 4 above)
Comment (haraldh) [reply tests/bin/bind.c:22]: This is an abstract unix socket without a file in the file system.

File: tests/bin/bind.c (hunk 5 above)
Comment (haraldh) [reply tests/bin/bind.c:24]: sa_len is only the path length

File: tests/bin/listen.c (hunk 6 above)
Comment (haraldh) [reply tests/bin/listen.c:26]: same

**Pull request merged**